README.md
# @endemolshinegroup/serverless-dynamodb-autoscaler

[![MIT Licensed][icon-license]][link-license]
[![NPM Version][icon-npm]][link-npm]
[![Build Status][icon-ci]][link-ci]
[![Greenkeeper Status][icon-greenkeeper]][link-greenkeeper]

[![Code Issues][icon-issues]][link-issues]
[![Codebase Maintainability][icon-maintainability]][link-maintainability]
[![Test Coverage][icon-coverage]][link-coverage]
[![Jest][icon-jest]][link-jest]

[![Commitizen][icon-commitizen]][link-commitizen]
[![Semantic Release][icon-semantic-release]][link-semantic-release]
[![Prettier][icon-prettier]][link-prettier]

Serverless Plugin for Amazon DynamoDB Auto Scaling

## Installation

```bash
yarn add -D @endemolshinegroup/serverless-dynamodb-autoscaler
```

## Usage

Add the plugin to your `serverless.yml`:

```yaml
plugins:
  - @endemolshinegroup/serverless-dynamodb-autoscaler
```

[icon-license]: https://img.shields.io/github/license/EndemolShineGroup/serverless-dynamodb-autoscaler.svg?longCache=true&style=flat-square
[link-license]: LICENSE
[icon-npm]: https://img.shields.io/npm/v/@endemolshinegroup/serverless-dynamodb-autoscaler.svg?longCache=true&style=flat-square
[link-npm]: https://www.npmjs.com/package/@endemolshinegroup/serverless-dynamodb-autoscaler
[icon-ci]: https://img.shields.io/travis/com/EndemolShineGroup/serverless-dynamodb-autoscaler.svg?longCache=true&style=flat-square
[link-ci]: https://travis-ci.com/EndemolShineGroup/serverless-dynamodb-autoscaler
[icon-greenkeeper]: https://img.shields.io/badge/greenkeeper-enabled-brightgreen.svg?longCache=true&style=flat-square
[link-greenkeeper]: https://greenkeeper.io/

[icon-issues]: https://img.shields.io/codeclimate/issues/EndemolShineGroup/serverless-dynamodb-autoscaler.svg?longCache=true&style=flat-square
[link-issues]: https://codeclimate.com/github/EndemolShineGroup/serverless-dynamodb-autoscaler/issues
[icon-maintainability]: https://img.shields.io/codeclimate/maintainability/EndemolShineGroup/serverless-dynamodb-autoscaler.svg?longCache=true&style=flat-square
[link-maintainability]: https://codeclimate.com/github/EndemolShineGroup/serverless-dynamodb-autoscaler
[icon-coverage]: https://img.shields.io/codecov/c/github/EndemolShineGroup/serverless-dynamodb-autoscaler/develop.svg?longCache=true&style=flat-square
[link-coverage]: https://codecov.io/gh/EndemolShineGroup/serverless-dynamodb-autoscaler

[icon-jest]: https://img.shields.io/badge/tested_with-jest-99424f.svg?longCache=true&style=flat-square
[link-jest]: https://jestjs.io/

[icon-commitizen]: https://img.shields.io/badge/commitizen-friendly-brightgreen.svg?longCache=true&style=flat-square
[link-commitizen]: http://commitizen.github.io/cz-cli/
[icon-semantic-release]: https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg?longCache=true&style=flat-square
[link-semantic-release]: https://semantic-release.gitbooks.io/semantic-release/
[icon-prettier]: https://img.shields.io/badge/code_style-prettier-ff69b4.svg?longCache=true&style=flat-square
[link-prettier]: https://prettier.io/



CHANGELOG.md
# 1.0.0 (2018-10-16)


### Bug Fixes

* **validion:** validation fix and tests improvement ([5d0a06f](https://github.com/EndemolShineGroup/serverless-dynamodb-autoscaler/commit/5d0a06f))


src/DynamoDBAutoscaling.spec.ts
import util from 'util';

import { cloneDeep } from 'lodash';

import Plugin from './DynamoDBAutoscaling';
import * as OutputMessages from './utils/output-messages';

import { expectedTemplate } from './__fixtures__/expectedCloudformationTemplate';

const serverless: any = {
  cli: {
    log: console.log,
  },
  service: {
    provider: {
      name: 'aws',
      compiledCloudFormationTemplate: {
        Resources: {},
      },
    },
    custom: {
      capacities: [
        {
          table: 'Table1',
          index: ['table1-index'],
          read: {
            minimum: 1,
            maximum: 5,
            usage: 0.2,
          },
          write: {
            minimum: 1,
            maximum: 5,
            usage: 0.2,
          },
        },
        {
          table: 'Table2',
          read: {
            minimum: 2,
            maximum: 10,
            usage: 0.8,
            scaleIn: 90,
          },
          write: {
            minimum: 4,
            maximum: 20,
            usage: 0.4,
            scaleOut: 75,
          },
        },
        {
          table: 'Table3',
          indexOnly: true,
          index: ['table3-index'],
          read: {
            minimum: 1,
            maximum: 5,
            usage: 0.2,
          },
          write: {
            minimum: 1,
            maximum: 5,
            usage: 0.2,
          },
        },
      ],
    },
  },
};

const getPlugingPromise = (config: any): Promise<any> => {
  return new Plugin({
    cli: {
      log: console.log,
    },
    ...config,
  }).hooks['package:compileEvents']();
};

describe('DynamoDBAutoscaling Plugin', () => {
  it('Table2 write should have default values set', () => {
    const data = cloneDeep(serverless);
    data.service.custom.capacities[1].write = {};
    return getPlugingPromise(data).then(() => {
      let targetObject: any = null;

      data.service.provider.compiledCloudFormationTemplate.Resources.AutoScalingPlan.Properties.ScalingInstructions.forEach(
        (instruction: any) => {
          if (
            instruction.ResourceId.payload[1][1]['Ref'] === 'Table2' &&
            instruction.ScalableDimension ===
              'dynamodb:table:WriteCapacityUnits'
          ) {
            targetObject = instruction;
          }
        },
      );

      expect(targetObject.MinCapacity).toEqual(5);
      expect(targetObject.MaxCapacity).toEqual(200);
      const tracking: any = targetObject.TargetTrackingConfigurations[0];
      expect(tracking.TargetValue).toEqual(75);
      expect(tracking.ScaleInCooldown).toEqual(60);
      expect(tracking.ScaleOutCooldown).toEqual(60);
    });
  });

  it('generate a valid template', () => {
    return getPlugingPromise(serverless).then(() => {
      expect(JSON.stringify(serverless, null, 2)).toEqual(expectedTemplate);
    });
  });
}); // End describe

describe('DynamoDBAutoscaling Plugin Errors', () => {
  let spy: any;

  beforeEach(() => {
    spy = jest.spyOn(console, 'log');
  });

  /*
    TODO?: Jest does not seem to like testing the promises the way I want (return expect(promise).rejects.toThrow());
    Therefore I'm using the hacky solution you will encounter in the tests below. We are testing for errors, so if the promise
    falls into `then` we have to fail the test. Any other tests you want to do (check error message and such) can be done inside
    `catch`. Any improvements to this is very welcome!
  */
  it('Empty configuration (Bad config)', () => {
    return getPlugingPromise({})
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalledWith(
          util.format(
            OutputMessages.CLI_SKIP,
            OutputMessages.INVALID_CONFIGURATION,
          ),
        );
      });
  });

  it('Partial configuration 1 (Bad config)', () => {
    return getPlugingPromise({
      service: {},
    })
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalledWith(
          util.format(
            OutputMessages.CLI_SKIP,
            OutputMessages.INVALID_CONFIGURATION,
          ),
        );
      });
  });

  it('Partial configuration 2 (Bad config)', () => {
    return getPlugingPromise({
      service: {
        provider: {},
      },
    })
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalledWith(
          util.format(
            OutputMessages.CLI_SKIP,
            OutputMessages.INVALID_CONFIGURATION,
          ),
        );
      });
  });

  it('Wrong provider "azure" (Bad config)', () => {
    return getPlugingPromise({
      service: {
        provider: {
          name: 'azure',
          compiledCloudFormationTemplate: {
            Resources: {},
          },
        },
      },
    })
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalledWith(
          util.format(OutputMessages.CLI_SKIP, OutputMessages.ONLY_AWS_SUPPORT),
        );
      });
  });

  it('Missing no custom (Bad config)', () => {
    return getPlugingPromise({
      service: {
        provider: {
          name: 'aws',
          compiledCloudFormationTemplate: {
            Resources: {},
          },
        },
      },
    })
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalledWith(
          util.format(
            OutputMessages.CLI_SKIP,
            OutputMessages.NO_AUTOSCALING_CONFIG,
          ),
        );
      });
  });

  it('Missing pluging config (Bad config)', () => {
    return getPlugingPromise({
      service: {
        provider: {
          name: 'aws',
          compiledCloudFormationTemplate: {
            Resources: {},
          },
        },
        custom: {},
      },
    })
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalledWith(
          util.format(
            OutputMessages.CLI_SKIP,
            OutputMessages.NO_AUTOSCALING_CONFIG,
          ),
        );
      });
  });

  it('pluging config is empty (Bad capacities config)', () => {
    return getPlugingPromise({
      service: {
        provider: {
          name: 'aws',
          compiledCloudFormationTemplate: {
            Resources: {},
          },
        },
        custom: {
          capacities: [],
        },
      },
    })
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalledWith(
          util.format(
            OutputMessages.CLI_SKIP,
            OutputMessages.NO_AUTOSCALING_CONFIG,
          ),
        );
      });
  });

  it('capacities config has only table', () => {
    return getPlugingPromise({
      service: {
        provider: {
          name: 'aws',
          compiledCloudFormationTemplate: {
            Resources: {},
          },
        },
        custom: {
          capacities: [
            {
              table: 'Table1',
            },
          ],
        },
      },
    })
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalled();
        const latestLog = spy.mock.calls[spy.mock.calls.length - 1][0];
        expect(latestLog).toMatch(
          new RegExp(OutputMessages.INVALID_TARGET_CONFIG),
        );
        expect(latestLog).toMatch(
          new RegExp(OutputMessages.MUST_DEFINE_READ_OR_WRITE),
        );
      });
  });

  it('capacities target minimum too low', () => {
    return getPlugingPromise({
      service: {
        provider: {
          name: 'aws',
          compiledCloudFormationTemplate: {
            Resources: {},
          },
        },
        custom: {
          capacities: [
            {
              table: 'Table1',
              write: {
                minimum: 0,
              },
            },
          ],
        },
      },
    })
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalled();
        const latestLog = spy.mock.calls[spy.mock.calls.length - 1][0];
        expect(latestLog).toMatch(
          new RegExp(OutputMessages.INVALID_TARGET_CONFIG),
        );
        expect(latestLog).toMatch(
          new RegExp(OutputMessages.CAPACITY_OUT_OF_BOUNDS),
        );
      });
  });

  it('capacities target minimum too high', () => {
    return getPlugingPromise({
      service: {
        provider: {
          name: 'aws',
          compiledCloudFormationTemplate: {
            Resources: {},
          },
        },
        custom: {
          capacities: [
            {
              table: 'Table1',
              write: {
                minimum: 1,
                maximum: 41000,
              },
            },
          ],
        },
      },
    })
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalled();
        const latestLog = spy.mock.calls[spy.mock.calls.length - 1][0];
        expect(latestLog).toMatch(
          new RegExp(OutputMessages.INVALID_TARGET_CONFIG),
        );
        expect(latestLog).toMatch(
          new RegExp(OutputMessages.CAPACITY_OUT_OF_BOUNDS),
        );
      });
  });

  it('capacities target usage out of bounds', () => {
    return getPlugingPromise({
      service: {
        provider: {
          name: 'aws',
          compiledCloudFormationTemplate: {
            Resources: {},
          },
        },
        custom: {
          capacities: [
            {
              table: 'Table1',
              write: {
                minimum: 1,
                maximum: 5,
                usage: 10,
              },
            },
          ],
        },
      },
    })
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalled();
        const latestLog = spy.mock.calls[spy.mock.calls.length - 1][0];
        expect(latestLog).toMatch(
          new RegExp(OutputMessages.INVALID_TARGET_CONFIG),
        );
        expect(latestLog).toMatch(
          new RegExp(OutputMessages.USAGE_OUT_OF_BOUNDS),
        );
      });
  });

  it('capacities target scaleIn not Int', () => {
    return getPlugingPromise({
      service: {
        provider: {
          name: 'aws',
          compiledCloudFormationTemplate: {
            Resources: {},
          },
        },
        custom: {
          capacities: [
            {
              table: 'Table1',
              write: {
                minimum: 1,
                maximum: 5,
                usage: 0.5,
                scaleIn: 0.1,
              },
            },
          ],
        },
      },
    })
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalled();
        const latestLog = spy.mock.calls[spy.mock.calls.length - 1][0];
        expect(latestLog).toMatch(
          new RegExp(OutputMessages.INVALID_TARGET_CONFIG),
        );
        expect(latestLog).toMatch(
          new RegExp(OutputMessages.SCALE_MUST_BE_INTEGER),
        );
      });
  });

  it('capacities target scaleIn not Int', () => {
    return getPlugingPromise({
      service: {
        provider: {
          name: 'aws',
          compiledCloudFormationTemplate: {
            Resources: {},
          },
        },
        custom: {
          capacities: [
            {
              table: 'Table1',
              write: {
                minimum: 1,
                maximum: 5,
                usage: 0.5,
                scaleIn: 1,
                scaleOut: 0.1,
              },
            },
          ],
        },
      },
    })
      .then(() => {
        //Force fail the test
        expect(true).toBeFalsy();
      })
      .catch(() => {
        expect(spy).toHaveBeenCalled();
        const latestLog = spy.mock.calls[spy.mock.calls.length - 1][0];
        expect(latestLog).toMatch(
          new RegExp(OutputMessages.INVALID_TARGET_CONFIG),
        );
        expect(latestLog).toMatch(
          new RegExp(OutputMessages.SCALE_MUST_BE_INTEGER),
        );
      });
  });
}); // End describe


src/utils/validation.ts
import { isArray } from 'lodash';

import { CapacityConfig, CapacityTarget } from '../types/plugin';
import * as OutputMessages from './output-messages';

const AWS_PROVIDER = 'aws';

let configErrors: string[] = [];

const validateInput = (input: any): Promise<string> => {
  configErrors = [];

  return new Promise((resolve, reject) => {
    // Check basic serverless template input
    if (!input) {
      return reject(errorMessage(OutputMessages.INVALID_CONFIGURATION));
    } else if (!input.service) {
      return reject(errorMessage(OutputMessages.INVALID_CONFIGURATION));
    } else if (!input.service.provider) {
      return reject(errorMessage(OutputMessages.INVALID_CONFIGURATION));
    } else if (!input.service.provider.name) {
      return reject(errorMessage(OutputMessages.INVALID_CONFIGURATION));
    } else if (input.service.provider.name !== AWS_PROVIDER) {
      return reject(errorMessage(OutputMessages.ONLY_AWS_SUPPORT));
    } else if (!input.service.custom) {
      return reject(errorMessage(OutputMessages.NO_AUTOSCALING_CONFIG));
    } else if (!input.service.custom.capacities) {
      return reject(errorMessage(OutputMessages.NO_AUTOSCALING_CONFIG));
    } else if (
      (input.service.custom.capacities &&
        !isArray(input.service.custom.capacities)) ||
      input.service.custom.capacities.length <= 0
    ) {
      return reject(errorMessage(OutputMessages.NO_AUTOSCALING_CONFIG));
    }
    // Check pluging config input
    if (!validateCapacityConfigs(input)) {
      return reject(errorMessage(getBadConfigErrorMessage()));
    }

    return resolve('');
  });
};

const errorMessage = (message: string) => {
  return { message };
};

const validateCapacityConfigs = (input: any): boolean => {
  const capacities = input.service.custom.capacities;

  let isTableFirstError: boolean = true;

  const isError = (tableName: string, message: string) => {
    if (isTableFirstError) {
      configErrors.push(`Errors found in ${tableName}:`);
    }
    isTableFirstError = false;
    configErrors.push(`  - ${message}`);
  };

  capacities.map((config: CapacityConfig) => {
    const targets: CapacityTarget[] = [];

    isTableFirstError = true;

    if (config.write) {
      targets.push(config.write);
    }

    if (config.read) {
      targets.push(config.read);
    }

    if (targets.length === 0) {
      isError(config.table, OutputMessages.MUST_DEFINE_READ_OR_WRITE);
    }

    targets.forEach((target: CapacityTarget) => {
      if (target.scaleIn && !Number.isInteger(target.scaleIn)) {
        isError(config.table, OutputMessages.SCALE_MUST_BE_INTEGER);
      }
      if (target.scaleOut && !Number.isInteger(target.scaleOut)) {
        isError(config.table, OutputMessages.SCALE_MUST_BE_INTEGER);
      }
      if (target.usage < 0.2 || target.usage > 0.8) {
        isError(config.table, OutputMessages.USAGE_OUT_OF_BOUNDS);
      }
      if (target.minimum < 1 || target.maximum > 40000) {
        isError(config.table, OutputMessages.CAPACITY_OUT_OF_BOUNDS);
      }
    });
  });

  return configErrors.length === 0;
};

const getBadConfigErrorMessage = (): string => {
  let errorMessage = OutputMessages.INVALID_TARGET_CONFIG;

  configErrors.forEach((error: string) => {
    errorMessage += '\n';
    errorMessage += error;
  });

  return errorMessage;
};

export default validateInput;


src/DynamoDBAutoscaling.ts
import util from 'util';

import { AutoScalingPlans, Refs } from 'cloudform';
import { isArray, merge } from 'lodash';

import getScalingInstruction from './getScalingInstruction';
import { CapacityConfig } from './types/plugin';
import * as OutputMessages from './utils/output-messages';
import validateInput from './utils/validation';

export default class DynamoDBAutoscaling {
  public hooks: {
    [key: string]: () => Promise<any>;
  };

  private serverless: any;

  constructor(serverless: any) {
    this.serverless = serverless;

    this.hooks = {
      'package:compileEvents': this.execute,
    };
  }

  private execute = (): Promise<any> => {
    return Promise.resolve()
      .then(() => {
        return validateInput(this.serverless);
      })
      .then(() => {
        return this.serverless.cli.log(util.format(OutputMessages.CLI_START));
      })
      .then(this.process)
      .then(() => {
        return this.serverless.cli.log(util.format(OutputMessages.CLI_DONE));
      })
      .catch((err: Error) => {
        this.serverless.cli.log(
          util.format(OutputMessages.CLI_SKIP, err.message),
        );
        throw new Error('aws-dynamodb-autoscaling failed with exit code 1');
      });
  };

  private process = () => {
    let scalingInstructions: any[] = [];
    const DependsOn: string[] = [];

    this.serverless.service.custom.capacities.map((config: CapacityConfig) => {
      if (config.index) {
        config.index = isArray(config.index) ? config.index : [config.index];
      }

      scalingInstructions = [
        ...scalingInstructions,
        ...this.getScalingInstructions(config),
      ];

      DependsOn.push(config.table);
    });

    const AutoScalingPlan = new AutoScalingPlans.ScalingPlan({
      ApplicationSource: new AutoScalingPlans.ScalingPlan.ApplicationSource({
        CloudFormationStackARN: Refs.StackId,
      }),
      ScalingInstructions: scalingInstructions,
    });

    AutoScalingPlan.DependsOn = DependsOn;

    merge(
      this.serverless.service.provider.compiledCloudFormationTemplate.Resources,
      { AutoScalingPlan },
    );
  };

  private getScalingInstructions = (config: CapacityConfig) => {
    let scalingInstructions: any[] = [];

    if (config.index) {
      // Create scaling instruction for each index
      (config.index as string[]).forEach((i: string) => {
        scalingInstructions = [
          ...scalingInstructions,
          ...this.generateInstructions(config),
        ];
      });

      if (config.indexOnly) {
        return scalingInstructions;
      }
    }

    // Create scaling instruction for the table
    scalingInstructions = [
      ...scalingInstructions,
      ...this.generateInstructions({
        table: config.table,
        write: config.write,
        read: config.read,
      }),
    ];

    return scalingInstructions;
  };

  private generateInstructions = (config: CapacityConfig) => {
    const scalingInstructions: any[] = [];

    this.serverless.cli.log(
      util.format(
        OutputMessages.CLI_RESOURCE,
        config.table,
        config.index ? '/index/' + config.index : '',
      ),
    );

    // Only add Auto Scaling for read capacity if configuration set is available
    if (!!config.read) {
      scalingInstructions.push(getScalingInstruction(config, true));
    }

    // Only add Auto Scaling for write capacity if configuration set is available
    if (!!config.write) {
      scalingInstructions.push(getScalingInstruction(config, false));
    }

    return scalingInstructions;
  };
}


src/getScalingInstruction.ts
import { AutoScalingPlans, Fn } from 'cloudform';

import { CapacityConfig } from './types/plugin';

const getScalingInstruction = (config: CapacityConfig, isRead: boolean) => {
  const { table, index } = config;

  const resource = ['table/', { Ref: table }];

  let type = 'table';

  if (index) {
    type = index as string;
    resource.push('/index/', index as string);
  }

  const target = isRead ? config.read! : config.write!;

  return new AutoScalingPlans.ScalingPlan.ScalingInstruction({
    ServiceNamespace: 'dynamodb',
    ResourceId: Fn.Join('', resource),
    MinCapacity: target.minimum ? target.minimum : 5,
    MaxCapacity: target.maximum ? target.maximum : 200,
    ScalableDimension: `dynamodb:${type}:${
      isRead ? 'Read' : 'Write'
    }CapacityUnits`,
    TargetTrackingConfigurations: [
      new AutoScalingPlans.ScalingPlan.TargetTrackingConfiguration({
        TargetValue: target.usage ? target.usage * 100 : 75,
        ScaleInCooldown: target.scaleIn ? target.scaleIn : 60,
        ScaleOutCooldown: target.scaleOut ? target.scaleOut : 60,
        PredefinedScalingMetricSpecification: new AutoScalingPlans.ScalingPlan.PredefinedScalingMetricSpecification(
          {
            PredefinedScalingMetricType: isRead
              ? 'DynamoDBReadCapacityUtilization'
              : 'DynamoDBWriteCapacityUtilization',
          },
        ),
      }),
    ],
  });
};

export default getScalingInstruction;


src/utils/output-messages.ts
export const CLI_DONE: string = `Added DynamoDB Auto Scaling to CloudFormation!`;
export const CLI_RESOURCE: string = ` - Building configuration for resource "table/%s%s"`;
export const CLI_SKIP: string = `Skipping DynamoDB Auto Scaling => %s!`;
export const CLI_START: string = `Configure DynamoDB Auto Scaling …`;
export const INVALID_CONFIGURATION: string = `Invalid serverless configuration`;
export const NO_AUTOSCALING_CONFIG: string = `Not Auto Scaling configuration found`;
export const ONLY_AWS_SUPPORT: string = `Only supported for AWS provider`;
export const SCALE_MUST_BE_INTEGER: string = `"ScaleIn" and "ScaleOut" must be integer`;
export const USAGE_OUT_OF_BOUNDS: string = `"usage" must be a value between 0.2 and 0.8`;
export const CAPACITY_OUT_OF_BOUNDS: string = `"minimum" must be greater than 0 and "maximum" must be 40000 or less`;
export const INVALID_TARGET_CONFIG: string = `Invalid capacities configuration`;
export const MUST_DEFINE_READ_OR_WRITE: string = `Missing definition for "write" and "read". At least 1 is needed`;


src/__fixtures__/expectedCloudformationTemplate.ts
export const expectedTemplate: string = `{
  "cli": {},
  "service": {
    "provider": {
      "name": "aws",
      "compiledCloudFormationTemplate": {
        "Resources": {
          "AutoScalingPlan": {
            "Type": "AWS::AutoScalingPlans::ScalingPlan",
            "Properties": {
              "ApplicationSource": {
                "CloudFormationStackARN": {
                  "Ref": "AWS::StackId"
                }
              },
              "ScalingInstructions": [
                {
                  "ServiceNamespace": "dynamodb",
                  "ResourceId": {
                    "Fn::Join": [
                      "",
                      [
                        "table/",
                        {
                          "Ref": "Table1"
                        },
                        "/index/",
                        [
                          "table1-index"
                        ]
                      ]
                    ]
                  },
                  "MinCapacity": 1,
                  "MaxCapacity": 5,
                  "ScalableDimension": "dynamodb:table1-index:ReadCapacityUnits",
                  "TargetTrackingConfigurations": [
                    {
                      "TargetValue": 20,
                      "ScaleInCooldown": 60,
                      "ScaleOutCooldown": 60,
                      "PredefinedScalingMetricSpecification": {
                        "PredefinedScalingMetricType": "DynamoDBReadCapacityUtilization"
                      }
                    }
                  ]
                },
                {
                  "ServiceNamespace": "dynamodb",
                  "ResourceId": {
                    "Fn::Join": [
                      "",
                      [
                        "table/",
                        {
                          "Ref": "Table1"
                        },
                        "/index/",
                        [
                          "table1-index"
                        ]
                      ]
                    ]
                  },
                  "MinCapacity": 1,
                  "MaxCapacity": 5,
                  "ScalableDimension": "dynamodb:table1-index:WriteCapacityUnits",
                  "TargetTrackingConfigurations": [
                    {
                      "TargetValue": 20,
                      "ScaleInCooldown": 60,
                      "ScaleOutCooldown": 60,
                      "PredefinedScalingMetricSpecification": {
                        "PredefinedScalingMetricType": "DynamoDBWriteCapacityUtilization"
                      }
                    }
                  ]
                },
                {
                  "ServiceNamespace": "dynamodb",
                  "ResourceId": {
                    "Fn::Join": [
                      "",
                      [
                        "table/",
                        {
                          "Ref": "Table1"
                        }
                      ]
                    ]
                  },
                  "MinCapacity": 1,
                  "MaxCapacity": 5,
                  "ScalableDimension": "dynamodb:table:ReadCapacityUnits",
                  "TargetTrackingConfigurations": [
                    {
                      "TargetValue": 20,
                      "ScaleInCooldown": 60,
                      "ScaleOutCooldown": 60,
                      "PredefinedScalingMetricSpecification": {
                        "PredefinedScalingMetricType": "DynamoDBReadCapacityUtilization"
                      }
                    }
                  ]
                },
                {
                  "ServiceNamespace": "dynamodb",
                  "ResourceId": {
                    "Fn::Join": [
                      "",
                      [
                        "table/",
                        {
                          "Ref": "Table1"
                        }
                      ]
                    ]
                  },
                  "MinCapacity": 1,
                  "MaxCapacity": 5,
                  "ScalableDimension": "dynamodb:table:WriteCapacityUnits",
                  "TargetTrackingConfigurations": [
                    {
                      "TargetValue": 20,
                      "ScaleInCooldown": 60,
                      "ScaleOutCooldown": 60,
                      "PredefinedScalingMetricSpecification": {
                        "PredefinedScalingMetricType": "DynamoDBWriteCapacityUtilization"
                      }
                    }
                  ]
                },
                {
                  "ServiceNamespace": "dynamodb",
                  "ResourceId": {
                    "Fn::Join": [
                      "",
                      [
                        "table/",
                        {
                          "Ref": "Table2"
                        }
                      ]
                    ]
                  },
                  "MinCapacity": 2,
                  "MaxCapacity": 10,
                  "ScalableDimension": "dynamodb:table:ReadCapacityUnits",
                  "TargetTrackingConfigurations": [
                    {
                      "TargetValue": 80,
                      "ScaleInCooldown": 90,
                      "ScaleOutCooldown": 60,
                      "PredefinedScalingMetricSpecification": {
                        "PredefinedScalingMetricType": "DynamoDBReadCapacityUtilization"
                      }
                    }
                  ]
                },
                {
                  "ServiceNamespace": "dynamodb",
                  "ResourceId": {
                    "Fn::Join": [
                      "",
                      [
                        "table/",
                        {
                          "Ref": "Table2"
                        }
                      ]
                    ]
                  },
                  "MinCapacity": 4,
                  "MaxCapacity": 20,
                  "ScalableDimension": "dynamodb:table:WriteCapacityUnits",
                  "TargetTrackingConfigurations": [
                    {
                      "TargetValue": 40,
                      "ScaleInCooldown": 60,
                      "ScaleOutCooldown": 75,
                      "PredefinedScalingMetricSpecification": {
                        "PredefinedScalingMetricType": "DynamoDBWriteCapacityUtilization"
                      }
                    }
                  ]
                },
                {
                  "ServiceNamespace": "dynamodb",
                  "ResourceId": {
                    "Fn::Join": [
                      "",
                      [
                        "table/",
                        {
                          "Ref": "Table3"
                        },
                        "/index/",
                        [
                          "table3-index"
                        ]
                      ]
                    ]
                  },
                  "MinCapacity": 1,
                  "MaxCapacity": 5,
                  "ScalableDimension": "dynamodb:table3-index:ReadCapacityUnits",
                  "TargetTrackingConfigurations": [
                    {
                      "TargetValue": 20,
                      "ScaleInCooldown": 60,
                      "ScaleOutCooldown": 60,
                      "PredefinedScalingMetricSpecification": {
                        "PredefinedScalingMetricType": "DynamoDBReadCapacityUtilization"
                      }
                    }
                  ]
                },
                {
                  "ServiceNamespace": "dynamodb",
                  "ResourceId": {
                    "Fn::Join": [
                      "",
                      [
                        "table/",
                        {
                          "Ref": "Table3"
                        },
                        "/index/",
                        [
                          "table3-index"
                        ]
                      ]
                    ]
                  },
                  "MinCapacity": 1,
                  "MaxCapacity": 5,
                  "ScalableDimension": "dynamodb:table3-index:WriteCapacityUnits",
                  "TargetTrackingConfigurations": [
                    {
                      "TargetValue": 20,
                      "ScaleInCooldown": 60,
                      "ScaleOutCooldown": 60,
                      "PredefinedScalingMetricSpecification": {
                        "PredefinedScalingMetricType": "DynamoDBWriteCapacityUtilization"
                      }
                    }
                  ]
                }
              ]
            },
            "DependsOn": [
              "Table1",
              "Table2",
              "Table3"
            ]
          }
        }
      }
    },
    "custom": {
      "capacities": [
        {
          "table": "Table1",
          "index": [
            "table1-index"
          ],
          "read": {
            "minimum": 1,
            "maximum": 5,
            "usage": 0.2
          },
          "write": {
            "minimum": 1,
            "maximum": 5,
            "usage": 0.2
          }
        },
        {
          "table": "Table2",
          "read": {
            "minimum": 2,
            "maximum": 10,
            "usage": 0.8,
            "scaleIn": 90
          },
          "write": {
            "minimum": 4,
            "maximum": 20,
            "usage": 0.4,
            "scaleOut": 75
          }
        },
        {
          "table": "Table3",
          "indexOnly": true,
          "index": [
            "table3-index"
          ],
          "read": {
            "minimum": 1,
            "maximum": 5,
            "usage": 0.2
          },
          "write": {
            "minimum": 1,
            "maximum": 5,
            "usage": 0.2
          }
        }
      ]
    }
  }
}`;
